### PR TITLE
feat: dedup for arrays

### PR DESCRIPTION
**Pull request by ermvrs**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/data_structures/src/array_ext.cairo [hunk 1]
```diff
@@ -21,6 +21,7 @@ trait ArrayTraitExt<T> {
     fn index_of_max<impl TPartialEq: PartialEq<T>, impl TPartialOrd: PartialOrd<T>>(
         self: @Array<T>
     ) -> Option<usize>;
+    fn dedup<impl TPartialEq: PartialEq<T>>(self: Array<T>) -> Array<T>;
```
Comment (gaetbout) [inline src/data_structures/src/array_ext.cairo:24]:
```suggestion
    fn dedup<impl TPartialEq: PartialEq<T>>(self: @Array<T>) -> Array<T>;
```

File: src/data_structures/src/array_ext.cairo [hunk 2]
```diff
@@ -135,6 +136,43 @@ impl ArrayImpl<T, impl TCopy: Copy<T>, impl TDrop: Drop<T>> of ArrayTraitExt<T>
...
+        }
+
+        let mut ret = array![];
+
+        let mut last_value = match self.pop_front() {
+            Option::Some(v) => v,
+            Option::None => {
+                return ret;
+            },
+        };
```
Comment (gaetbout) [inline src/data_structures/src/array_ext.cairo:154]: as you already made sure len >1 you can just pop_front().unwrap()

File: src/data_structures/src/array_ext.cairo [hunk 3]
```diff
@@ -135,6 +136,43 @@ impl ArrayImpl<T, impl TCopy: Copy<T>, impl TDrop: Drop<T>> of ArrayTraitExt<T>
...
+
+        loop {
+            match self.pop_front() {
+                Option::Some(v) => {
+
+                    if(last_value == v) {
+                        continue;
+                    };
+                    last_value = v;
+                    ret.append(v);
```
Comment (gaetbout) [inline src/data_structures/src/array_ext.cairo:166]: maybe just me but I'd just do:
```suggestion
                    if(last_value != v) {
                      last_value = v;
                      ret.append(v);
                    };
```

File: src/data_structures/src/tests/array_ext_test.cairo [hunk 4]
```diff
@@ -3,6 +3,111 @@ use array::{ArrayTrait, SpanTrait};
 
 use alexandria_data_structures::array_ext::{ArrayTraitExt, SpanTraitExt};
 
+// dedup
```
Comment (gaetbout) [inline src/data_structures/src/tests/array_ext_test.cairo:6]: Amazing tests :)

File: src/data_structures/src/array_ext.cairo (hunk 2 above)
Comment (ermvrs) [reply src/data_structures/src/array_ext.cairo]: Done

File: src/data_structures/src/array_ext.cairo [hunk 5]
```diff
@@ -135,6 +136,37 @@ impl ArrayImpl<T, impl TCopy: Copy<T>, impl TDrop: Drop<T>> of ArrayTraitExt<T>
     ) -> Option<usize> {
         self.span().index_of_max()
     }
+
+    fn dedup<impl TPartialEq: PartialEq<T>>(mut self: Array<T>) -> Array<T> {
+        let len = self.len();
+
+        if(len <= 1) {
+            return self;
+        }
```
Comment (gaetbout) [inline src/data_structures/src/array_ext.cairo:145]:
```suggestion
        if( self.len() <= 1) {
            return self;
        }
```

**Comment by gaetbout** [comment]
Pinging the chief @LucasLvy 🙏

File: src/data_structures/src/array_ext.cairo [hunk 6]
```diff
@@ -135,6 +136,35 @@ impl ArrayImpl<T, impl TCopy: Copy<T>, impl TDrop: Drop<T>> of ArrayTraitExt<T>
...
+        let mut ret = array![];
+
+        let mut last_value = self.pop_front().unwrap();
+
+        ret.append(last_value);
+
+        loop {
+            match self.pop_front() {
+                Option::Some(v) => {
+
```
Comment (gaetbout) [inline src/data_structures/src/array_ext.cairo:154]: Maybe get rid of that extra line, but I'm getting picky now 😄

Comment (ermvrs) [reply src/data_structures/src/array_ext.cairo:154]: Line: 154 destroyed 💣

File: src/data_structures/src/array_ext.cairo [hunk 7]
```diff
@@ -135,6 +136,33 @@ impl ArrayImpl<T, impl TCopy: Copy<T>, impl TDrop: Drop<T>> of ArrayTraitExt<T>
...
+    fn dedup<impl TPartialEq: PartialEq<T>>(mut self: Array<T>) -> Array<T> {
+        if(self.len() <= 1) {
+            return self;
+        }
+
+        let mut ret = array![];
+        let mut last_value = self.pop_front().unwrap();
+
+        ret.append(last_value);
```
Comment (LucasLvy) [inline src/data_structures/src/array_ext.cairo:148]:
```suggestion    
        let mut last_value = self.pop_front().unwrap();
        let mut ret = array![last_value];
```

File: src/data_structures/src/array_ext.cairo [hunk 8]
```diff
@@ -135,6 +136,33 @@ impl ArrayImpl<T, impl TCopy: Copy<T>, impl TDrop: Drop<T>> of ArrayTraitExt<T>
...
+
+        loop {
+            match self.pop_front() {
+                Option::Some(v) => {
+                    if(last_value != v) {
+                      last_value = v;
+                      ret.append(v);
+                    };
+                },
+                Option::None(()) => {
```
Comment (LucasLvy) [inline src/data_structures/src/array_ext.cairo:158]:
```suggestion
                Option::None => {
```

File: src/data_structures/src/array_ext.cairo [hunk 9]
```diff
@@ -135,6 +136,33 @@ impl ArrayImpl<T, impl TCopy: Copy<T>, impl TDrop: Drop<T>> of ArrayTraitExt<T>
...
+
+        ret.append(last_value);
+
+        loop {
+            match self.pop_front() {
+                Option::Some(v) => {
+                    if(last_value != v) {
+                      last_value = v;
+                      ret.append(v);
+                    };
```
Comment (LucasLvy) [inline src/data_structures/src/array_ext.cairo:156]: This is wrong

**Review by LucasLvy — CHANGES_REQUESTED** [review]
Please:
- Format
- Add a test with `[1, 2, 1]` for example and check that the length is 2.
- The logic seems wrong here you're just checking the next value is not the same as the current one which is not what we would expect from remove duplicates. You'll need to iterate through the array everytime to check if it's already there or not.

**Comment by gaetbout** [comment]
Any reason you reverted the use of @Array ?

**Comment by ermvrs** [comment]
Otherwise i need to clone the array to return at the beginning and use pop_front. Will use more gas :(

**Comment by ermvrs** [comment]
@gaetbout Return type is Array<T> and I also need to use pop_front inside function. So to return will have to same type as parameter in first if case (len < 1). If parameter type is @array then I have to clone it to new array to return that one in first case.
Although further it can be changed if we implement same methods for span as well.

File: src/data_structures/src/array_ext.cairo [hunk 10]
```diff
@@ -135,6 +136,31 @@ impl ArrayImpl<T, impl TCopy: Copy<T>, impl TDrop: Drop<T>> of ArrayTraitExt<T>
     ) -> Option<usize> {
         self.span().index_of_max()
     }
+
+    fn dedup<impl TPartialEq: PartialEq<T>>(mut self: Array<T>) -> Array<T> {
+        if (self.len() <= 1) {
```
Comment (gaetbout) [inline src/data_structures/src/array_ext.cairo:141]:
```suggestion
        if self.len() <= 1 {
```

**Comment by gaetbout** [comment]
You could also do in array_ext
```rust
    fn dedup<impl TPartialEq: PartialEq<T>>(mut self: @Array<T>) -> Array<T> {
        self.span().dedup()
    }
```
And add this to span_ext
```rust
    fn dedup<impl TPartialEq: PartialEq<T>>(mut self: Span<T>) -> Array<T> {
        if self.len() == 0 {
            return array![];
        }
        if (self.len() <= 1) {
            return array![*self[0]];
        }

        let mut last_value = self.pop_front().unwrap();
        let mut ret = array![*last_value];

        loop {
            match self.pop_front() {
                Option::Some(v) => {
                    if (last_value != v) {
                        last_value = v;
                        ret.append(*v);
                    };
                },
                Option::None => {
                    break;
                }
            };
        };

        ret
    }
```

**Comment by ermvrs** [comment]
Done, changed like your suggestion

File: src/data_structures/src/array_ext.cairo [hunk 11]
```diff
@@ -354,4 +360,32 @@ impl SpanImpl<T, impl TCopy: Copy<T>, impl TDrop: Drop<T>> of SpanTraitExt<T> {
...
         }
     }
+
+    fn dedup<impl TPartialEq: PartialEq<T>>(mut self: Span<T>) -> Array<T> {
+        if self.len() == 0 {
+            return array![];
+        }
+        if (self.len() <= 1) {
+            return array![*self[0]];
+        }
```
Comment (gaetbout) [inline src/data_structures/src/array_ext.cairo:370]: Just realized that this is useless, think you can just removed that

File: src/data_structures/src/array_ext.cairo [hunk 12]
```diff
@@ -43,6 +44,7 @@ trait SpanTraitExt<T> {
     fn index_of_max<impl TPartialEq: PartialEq<T>, impl TPartialOrd: PartialOrd<T>>(
         self: Span<T>
     ) -> Option<usize>;
+    fn dedup<impl TPartialEq: PartialEq<T>>(self: Span<T>) -> Array<T>;
```
Comment (LucasLvy) [inline src/data_structures/src/array_ext.cairo:47]: can you add documentation on what the function does plz

**Review by LucasLvy — CHANGES_REQUESTED** [review]

File: src/data_structures/src/array_ext.cairo (hunk 12 above)
Comment (ermvrs) [reply src/data_structures/src/array_ext.cairo:47]: Where should i add that

Comment (LucasLvy) [reply src/data_structures/src/array_ext.cairo:47]: right above the function, rust style doc

**Review by LucasLvy — APPROVED** [review]

**Pull request merged**